### PR TITLE
Implement ```implicit``` for ```subroutine```

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes: #894

### CODE REVIEW

File: src/lfortran/semantics/ast_symboltable_visitor.cpp [hunk 1]
```diff
@@ -253,7 +253,115 @@ class SymbolTableVisitor : public CommonVisitor<SymbolTableVisitor> {
...
+        }
+
+        for (char ch='o'; ch<='z'; ch++) {
+            implicit_dictionary[std::string(1, ch)] = LFortran::ASRUtils::TYPE(ASR::make_Real_t(al, x.base.base.loc, 4, nullptr, 0));
+        }
+
+        for (char ch='a'; ch<='h'; ch++) {
+            implicit_dictionary[std::string(1, ch)] = LFortran::ASRUtils::TYPE(ASR::make_Real_t(al, x.base.base.loc, 4, nullptr, 0));
+        }
+    }
```
Comment (certik) [inline src/lfortran/semantics/ast_symboltable_visitor.cpp:268]: I think this is identical to the other function, so you can reuse it.

File: src/lfortran/semantics/ast_symboltable_visitor.cpp [hunk 2]
```diff
@@ -253,7 +253,115 @@ class SymbolTableVisitor : public CommonVisitor<SymbolTableVisitor> {
...
+                        implicit_dictionary[std::string(1, *end)] = type;
+                    } else {
+                        for(char ch=*start; ch<=*end; ch++){
+                            implicit_dictionary[std::string(1, ch)] = type;
+                        }
+                    }
+                }
+            }
+        }
+    }
```
Comment (certik) [inline src/lfortran/semantics/ast_symboltable_visitor.cpp:350]: I think this function is identical, so you should create a templated function and call it from both, to not duplicate code.

Comment (Pranavchiku) [reply src/lfortran/semantics/ast_symboltable_visitor.cpp:350]: Yeah, I was thinking the, same wil do that.

**Comment by Pranavchiku** [comment]
I guess it might fail ```CI / Documentation (pull_request) ```, it is because we must return something at the end of ```process_implicit_statements``` template function, I will look at it as soon as possible.

File: src/lfortran/semantics/ast_symboltable_visitor.cpp [hunk 3]
```diff
@@ -253,7 +253,127 @@ class SymbolTableVisitor : public CommonVisitor<SymbolTableVisitor> {
...
+        for (char ch='o'; ch<='z'; ch++) {
+            implicit_dictionary[std::string(1, ch)] = LFortran::ASRUtils::TYPE(ASR::make_Real_t(al, a_loc, 4, nullptr, 0));
+        }
+
+        for (char ch='a'; ch<='h'; ch++) {
+            implicit_dictionary[std::string(1, ch)] = LFortran::ASRUtils::TYPE(ASR::make_Real_t(al, a_loc, 4, nullptr, 0));
+        }
+    }
+
+	template <typename T> T process_implicit_statements(const T &x, std::map<std::string, ASR::ttype_t*> &implicit_dictionary) {
```
Comment (certik) [inline src/lfortran/semantics/ast_symboltable_visitor.cpp:270]:
```suggestion
	template <typename T> process_implicit_statements(const T &x, std::map<std::string, ASR::ttype_t*> &implicit_dictionary) {
```

Comment (certik) [reply src/lfortran/semantics/ast_symboltable_visitor.cpp:270]: This should fix the warning.

**Comment by Pranavchiku** [comment]
@certik this PR can have a final review.

File: src/lfortran/semantics/ast_symboltable_visitor.cpp [hunk 4]
```diff
@@ -253,7 +253,128 @@ class SymbolTableVisitor : public CommonVisitor<SymbolTableVisitor> {
...
+                        for(char ch=*start; ch<=*end; ch++){
+                            implicit_dictionary[std::string(1, ch)] = type;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+	void print_implicit_dictionary(std::map<std::string, ASR::ttype_t*> &implicit_dictionary) {
```
Comment (certik) [inline src/lfortran/semantics/ast_symboltable_visitor.cpp:353]: Inconsistent indenting....

File: src/lfortran/semantics/ast_symboltable_visitor.cpp [hunk 5]
```diff
@@ -253,7 +253,128 @@ class SymbolTableVisitor : public CommonVisitor<SymbolTableVisitor> {
...
+
+        for (char ch='a'; ch<='h'; ch++) {
+            implicit_dictionary[std::string(1, ch)] = LFortran::ASRUtils::TYPE(ASR::make_Real_t(al, a_loc, 4, nullptr, 0));
+        }
+    }
+
+	template <typename T> 
+	void process_implicit_statements(const T &x, std::map<std::string, ASR::ttype_t*> &implicit_dictionary) {
+        //iterate over all implicit statements
+        for (size_t i=0;i<x.n_implicit;i++) {
```
Comment (certik) [inline src/lfortran/semantics/ast_symboltable_visitor.cpp:273]: Use just 4 spaces for indenting, not 8.

File: tests/reference/asr-implicit11-71e42a3.stdout [hunk 6]
```diff
@@ -146,7 +146,7 @@
                                     () 
                                     () 
                                     Default 
-                                    (Real 4 []) 
+                                    (Complex 8 [])
```
Comment (certik) [inline tests/reference/asr-implicit11-71e42a3.stdout:149]: Why did this test change? Was it incorrect previously, or is it incorrect now? We need to investigate this.

**Comment by Pranavchiku** [comment]
So sorry for the indenting issues, there is some trouble with my vscode I will figure it out.

File: tests/reference/asr-implicit11-71e42a3.stdout (hunk 6 above)
Comment (Pranavchiku) [reply tests/reference/asr-implicit11-71e42a3.stdout:149]: will check that.

**Comment by certik** [comment]
> So sorry for the indenting issues, there is some trouble with my vscode I will figure it out.

Just manually and carefully review the actual PR diff. That's what I do for my own code and I often find things that I missed before.

File: tests/reference/asr-implicit11-71e42a3.stdout (hunk 6 above)
Comment (Pranavchiku) [reply tests/reference/asr-implicit11-71e42a3.stdout:149]: According to me, it is correct now as ```implicit complex*16 (z,i)``` must have z as ```complex 16```.

Comment (certik) [reply tests/reference/asr-implicit11-71e42a3.stdout:149]: Ok, but why was it showing as Real 4 before?

**Comment by Pranavchiku** [comment]
@certik this PR can have a final review.

File: src/lfortran/semantics/ast_symboltable_visitor.cpp [hunk 7]
```diff
@@ -355,6 +476,16 @@ class SymbolTableVisitor : public CommonVisitor<SymbolTableVisitor> {
            matter since we would have already checked the intent */
         current_procedure_args.clear();
         current_procedure_abi_type = ASR::abiType::Source;
+
+		// print_implicit_dictionary(implicit_dictionary);
```
Comment (certik) [inline src/lfortran/semantics/ast_symboltable_visitor.cpp:480]:
```suggestion
        // print_implicit_dictionary(implicit_dictionary);
```

**Pull request merged**